portfolio/src/main/webapp/script.js

```diff
@@ -34,30 +34,57 @@ function gotoRandomSite() {
 }
 
 function getComments() {
-    // Perform the fetch and store as promise
-    var commentsPromise = fetch('/data');
-
-    // Use callbacks on promise to extract the JSON object and put into HTML
-    commentsPromise.then(response => response.json()).then((resJson) => {
-        const commentsContainer = document.getElementById("comments-container");
-        commentsContainer.innerText = '';
-        
-        resJson.forEach(comment => {
-            commentsContainer.appendChild(createParagraphElement(comment));
-        })
+  // Perform the fetch and store as promise
+  var commentsPromise = fetch('/data');
+
+  // Use callbacks on promise to extract the JSON object and put into HTML
+  commentsPromise.then(response => response.json()).then((resJson) => {
+    const commentsContainer = document.getElementById("comments-container");
+    commentsContainer.innerText = '';
+
+    resJson.forEach(comment => {
+      commentsContainer.appendChild(createParagraphElement(comment));
+    })
+  });
+}
+
+function translateComments() {
+  var container = document.getElementById('comments-container');
```

Comment:
Extracting comments from comments-container introduces javascript assumption about HTML document structure (there is a comments-container element whose immediate children contain comments). If you change HTML to introduce additional heading or author for each comment then this assumption will be broken. 

Alternatively you could retain comments retrieved by getComments in some variable (this.comments = resJson) and use it here to extract raw comments in the form you received them from server. This way you could structure javascript into following functions: 
- Function that retrieves comments and stores them in a variabe
- Function that translates comments stored in that variable and updates them there
- Function that creates DOM representation of comments, stored in the variable. (createParagraphElement)

Reply:
Gotcha, this sounds like a good roadmap. I have a question, though. If I were to have that function that retrieves comments and store them, and then another to translate them and store them, they would both also have to update the DOM with the new comments *within* those functions right? Because they'd both be asynchronous and working with promises, if I took the responsibility of updating the DOM to a separate function only, I'd have a race condition right? Because at any point that function could try to update the DOM before the comments variable has been initialized?

Reply:
There are many alternatives and they depend on desired behavior. E.g.:
- load comments from onload, but have button to trigger translation
- Chain comments fetch and translation. Comments fetch returns a promise. Once you get the list of comments, instead of single 'then' that updates DOM you fork it into two steps: one updates DOM, another one sends comments to translate server and returns another promise. Once that promise is satisfied you update DOM with translated comments.
- Do everything on the server. When you fetch comments you pass language code. Servlet reads comments from datastore, translates them and returns translated versions.